File: daily/Binary Search/875-Koko-Eating-Bananas.go
package daily

import "math"

func minEatingSpeed(piles []int, h int) int {
	max := piles[0]
	for _, pile := range piles {
		if pile > max {
			max = pile
		}
	}

	left, right := 1, max
	ans := 0
	for left <= right {
		mid := (left + right) / 2
		if validSpeed(piles, h, mid) {
			ans = mid
			right = mid - 1
		} else {
			left = mid + 1
		}
	}
	return ans
}

func validSpeed(piles []int, h int, speed int) bool {
	hours := 0
	for _, pile := range piles {
		hours += int(math.Ceil(float64(pile) / float64(speed)))
		if hours > h {
			return false
		}
	}
	return hours <= h
}

File: daily/Binary Search/875-Koko-Eating-Bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        left, right = 1, max(piles)
        ans = 0
        while left <= right:
            mid = (left + right) // 2
            if self.validSpeed(piles, h, mid):
                ans = mid
                right = mid - 1
            else:
                left = mid + 1
        return ans
    def validSpeed(self, piles: List[int], h: int, speed: int) -> bool:
        hours = 0
        for pile in piles:
            hours += math.ceil(float(pile) / speed)
            if hours > h:
                return False
        return hours <= h